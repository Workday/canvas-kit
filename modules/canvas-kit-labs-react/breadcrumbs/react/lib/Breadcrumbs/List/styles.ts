import {CSSObject} from '@emotion/core';

// default max-width for truncating text
const DEFAULT_MAX_WIDTH = 350;

export const truncateStyles = (maxWidth: number = DEFAULT_MAX_WIDTH): CSSObject => ({
  display: 'inline-block',
  maxWidth,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
});
