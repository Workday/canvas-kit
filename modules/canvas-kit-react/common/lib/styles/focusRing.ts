import {keyframes, CSSObject} from '@emotion/core';
import canvas from '@workday/canvas-kit-react-core';
import {EmotionCanvasTheme, defaultCanvasTheme} from '../theming/index';
import memoize from 'lodash/memoize';

interface FocusRingOptions {
  width?: number;
  separation?: number;
  animate?: boolean;
  /**
   * Specifies where the ring(s) should be attached.
   * - undefined: Both "inner" and "outer" shadows outside the container.
   * - 'inner': "Inner" shadow inset. "Outer" shadow outside the container.
   * - 'outer': Both "inner" and "outer" shadows inside the container.
   */
  inset?: 'inner' | 'outer';
  innerColor?: string;
  outerColor?: string;
  memoize?: boolean;
}

function calculateFocusRing({
  width,
  separation,
  animate,
  inset,
  innerColor,
  outerColor,
}: Required<Omit<FocusRingOptions, 'memoize' | 'inset'>> &
  Pick<FocusRingOptions, 'inset'>): CSSObject {
  let boxShadow, innerWidth, outerWidth;

  switch (inset) {
    case 'outer':
      innerWidth = width + separation;
      outerWidth = separation;
      boxShadow = `inset 0 0 0 ${outerWidth}px ${outerColor}, inset 0 0 0 ${innerWidth}px ${innerColor}`;
      break;

    case 'inner':
      innerWidth = separation;
      outerWidth = width;
      boxShadow = `inset 0 0 0 ${innerWidth}px ${innerColor}, 0 0 0 ${outerWidth}px ${outerColor}`;
      break;

    default:
      innerWidth = separation;
      outerWidth = width + separation;
      boxShadow = `0 0 0 ${innerWidth}px ${innerColor}, 0 0 0 ${outerWidth}px ${outerColor}`;
      break;
  }

  if (animate) {
    const fadeIn = keyframes({
      '0%': {boxShadow},
      '100%': {boxShadow},
    });

    return {animation: `${fadeIn} 100ms`, boxShadow};
  }

  return {boxShadow};
}

export const memoizedFocusRing = memoize(calculateFocusRing, (...args) => JSON.stringify(args));

/**
 * A utility to create a canvas style focus ring around your widget.
 * By default, this mixin will create a 2px focus ring tightly wrapped
 * to the container (no whitespace).
 *
 * @param width        Allows the user to specify the thickness in px of the focus ring.
 * @param separation  Allows the user to define the width in px of the whitespace
 *                         that there should be between the component and the focus ring.
 * @param animate          Set property to false to opt out of the standard grow out of the middle animation
 * @param inset            Determines whether or not the focus ring is inset
 * @param innerShadowColor Allows the user to specify the inner shadow color
 * @param outerShadowColor Allows the user to specify the outer shadow color
 *
 * @returns {CSSObject} the css object for the focus ring style
 */

/**
 * A utility to create a canvas style focus ring around your widget.
 * By default, this mixin will create a 2px focus ring tightly wrapped
 * to the container (no whitespace).
 *
 * @returns {CSSObject} the css object for the focus ring style
 */
export function focusRing(options: FocusRingOptions = {}, theme?: EmotionCanvasTheme): CSSObject {
  const {
    width = 2,
    separation = 0,
    animate = true,
    innerColor = canvas.colors.frenchVanilla100,
    outerColor = theme && theme.canvas
      ? theme.canvas.palette.common.focusOutline
      : defaultCanvasTheme.palette.common.focusOutline,
    memoize = true,
    inset,
  } = options;

  const args = {
    width: width,
    separation: separation,
    innerColor: innerColor,
    outerColor: outerColor,
    animate,
    inset,
  };

  if (memoize) {
    return memoizedFocusRing(args);
  }

  return calculateFocusRing(args);
}
