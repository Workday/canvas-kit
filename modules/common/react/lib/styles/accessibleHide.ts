import {CSSObject} from '@emotion/core';

/**
 * A utility to hide content normal users while still making accessible to screenreaders
 * See https://a11y-101.com/development/skip-link
 */
export const accessibleHide: CSSObject = {
  clip: 'rect(1px, 1px, 1px, 1px)', // Deprecated but still used by most browsers, clip-path will be taking its place soon.
  clipPath: 'polygon(0px 0px, 0px 0px, 0px 0px, 0px 0px)',
  position: 'absolute',
  overflow: 'hidden',
  whiteSpace: 'nowrap',
  height: '1px',
  width: '1px',
  margin: '-1px',
  padding: 0,
  border: 0,
};
