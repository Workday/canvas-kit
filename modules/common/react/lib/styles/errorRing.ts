import {ErrorType} from '../types';
import {CSSObject} from 'create-emotion';
import {colors, inputColors} from '@workday/canvas-kit-react-core';

export default function errorRing(error?: ErrorType): CSSObject {
  let errorBorderColor;
  let errorRingColor;

  if (error === ErrorType.Error) {
    errorRingColor = inputColors.error.border;
  } else if (error === ErrorType.Alert) {
    errorRingColor = inputColors.warning.border;
    errorBorderColor = colors.cantaloupe600;
  } else {
    return {};
  }

  return {
    border: errorBorderColor ? `1px solid ${errorBorderColor}` : 'none',
    transition: '100ms box-shadow',
    boxShadow: `inset 0 0 0 2px ${errorRingColor}`,
    '&:hover': {
      borderColor: errorBorderColor,
    },
    '&:focus:not([disabled])': {
      borderColor: errorBorderColor,
      boxShadow: `inset 0 0 0 2px ${errorRingColor},
        0 0 0 2px ${colors.frenchVanilla100},
        0 0 0 4px ${inputColors.focusBorder}`,
    },
  };
}
