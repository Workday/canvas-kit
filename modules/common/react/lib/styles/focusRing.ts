import {keyframes, CSSObject} from '@emotion/core';
import canvas from '@workday/canvas-kit-react-core';
import {CanvasTheme, defaultCanvasTheme} from '../theming/index';
import memoize from 'lodash/memoize';

interface FocusRingOptions {
  width?: number;
  separation?: number;
  animate?: boolean;
  inset?: boolean;
  innerColor?: string;
  outerColor?: string;
  memoize?: boolean;
}

function calculateFocusRing({
  width,
  separation,
  animate,
  inset,
  innerColor,
  outerColor,
}: Omit<Required<FocusRingOptions>, 'memoize'>): CSSObject {
  const endingInnerShadow = (inset ? 'inset ' : '') + '0 0 0 ' + separation + 'px ' + innerColor;
  const endingOuterShadow =
    (inset ? 'inset ' : '') + '0 0 0 ' + (width + separation) + 'px ' + outerColor;
  const endingBoxShadow = inset
    ? `${endingOuterShadow}`
    : `${endingInnerShadow}, ${endingOuterShadow}`;

  if (animate) {
    const fadeIn = keyframes({
      '0%': {boxShadow: endingBoxShadow},
      '100%': {boxShadow: endingBoxShadow},
    });

    return {animation: `${fadeIn} 100ms`, boxShadow: endingBoxShadow};
  }

  return {boxShadow: endingBoxShadow};
}

export const memoizedFocusRing = memoize(calculateFocusRing, (...args) => JSON.stringify(args));

/**
 * A utility to create a canvas style focus ring around your widget.
 * By default, this mixin will create a 2px focus ring tightly wrapped
 * to the container (no whitespace).
 *
 * @param width        Allows the user to specify the thickness in px of the focus ring.
 * @param separation  Allows the user to define the width in px of the whitespace
 *                         that there should be between the component and the focus ring.
 * @param animate          Set property to false to opt out of the standard grow out of the middle animation
 * @param inset            Determines whether or not the focus ring is inset
 * @param innerShadowColor Allows the user to specify the inner shadow color
 * @param outerShadowColor Allows the user to specify the outer shadow color
 *
 * @returns {CSSObject} the css object for the focus ring style
 */

/**
 * A utility to create a canvas style focus ring around your widget.
 * By default, this mixin will create a 2px focus ring tightly wrapped
 * to the container (no whitespace).
 *
 * @returns {CSSObject} the css object for the focus ring style
 */
export function focusRing(
  options: FocusRingOptions = {},
  theme: CanvasTheme = defaultCanvasTheme
): CSSObject {
  const {
    width = 2,
    separation = 0,
    animate = true,
    inset = false,
    innerColor = canvas.colors.frenchVanilla100,
    outerColor = theme.palette.common.focusOutline,
    memoize = true,
  } = options;

  const args = {
    width: width,
    separation: separation,
    innerColor: innerColor,
    outerColor: outerColor,
    animate,
    inset,
  };

  if (memoize) {
    return memoizedFocusRing(args);
  }

  return calculateFocusRing(args);
}
