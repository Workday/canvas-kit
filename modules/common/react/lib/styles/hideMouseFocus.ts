import {CSSObject} from '@emotion/core';

/**
 * A utility to hide the default canvas style focus ring when using mouse input.
 * Requires wrapping your components in an InputProvider component.
 */
export const hideMouseFocus: CSSObject = {
  [`[data-whatinput='mouse'] &:focus,
    [data-whatinput='touch'] &:focus,
    [data-whatinput='pointer'] &:focus`]: {
    outline: 'none',
    boxShadow: 'none',
    animation: 'none',
  },
};

type IndexableObject = {[key: string]: object};

/*
 * A utility that simplifies focus selectors since you can't use nested syntax for some reason. Example:
 * [`[data-whatinput="mouse"],
     [data-whatinput="touch"],
     [data-whatinput="pointer"]: {
       '& .my-selector': {
          ...
       }
   },
 */
export const mouseFocusBehavior = (selectors: IndexableObject) => {
  const output: IndexableObject = {};

  Object.keys(selectors).map((selector, index) => {
    selector.split(',').forEach(selectorPart => {
      const prefixedSelector = `
        [data-whatinput="mouse"] ${selectorPart},
        [data-whatinput="touch"] ${selectorPart},
        [data-whatinput="pointer"] ${selectorPart}`;
      output[prefixedSelector] = selectors[selector];
    });
  });

  return output;
};
