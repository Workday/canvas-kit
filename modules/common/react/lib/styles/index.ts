export * from './accessibleHide';
export * from './errorRing';
export * from './focusRing';
export * from './hideMouseFocus';
