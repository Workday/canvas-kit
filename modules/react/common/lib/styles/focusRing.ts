import {brand, base} from '@workday/canvas-tokens-web';

/**
 * A utility to create a canvas style focus ring around your widget.
 * By default, this mixin will create a 2px focus ring tightly wrapped
 * to the container (no whitespace).
 */
export function focusRing<
  Width extends number = 2,
  Separation extends number = 2,
  InnerColor extends string = `var(${typeof base.frenchVanilla100})`,
  OuterColor extends string = `var(${typeof brand.common.focusOutline})`,
  Inset extends 'inner' | 'outer' | 'neither' = 'neither'
>(
  options: {
    /**
     * Allows the user to specify the thickness in px of the focus ring.
     */
    width?: Width;
    /**
     * Allows the user to define the width in px of the whitespace
     */
    separation?: Separation;
    /**
     * Allows the user to specify the inner shadow color
     */
    innerColor?: InnerColor;
    /**
     * Allows the user to specify the outer shadow color
     */
    outerColor?: OuterColor;
    /**
     * Determines whether or not the focus ring is inset. Specifies where the ring(s) should be
     * attached.
     * - undefined: Both "inner" and "outer" shadows outside the container.
     * - 'inner': "Inner" shadow inset. "Outer" shadow outside the container.
     * - 'outer': Both "inner" and "outer" shadows inside the container.
     */
    inset?: Inset;
  } = {}
): FocusRingReturn<Width, Separation, InnerColor, OuterColor, Inset> {
  const {
    width = 2,
    separation = 0,
    innerColor = `var(${base.frenchVanilla100})`,
    outerColor = `var(${brand.common.focusOutline})`,
    inset = 'neither',
  } = options;

  let boxShadow;

  switch (inset) {
    case 'outer':
      boxShadow = `inset 0 0 0 ${separation}px ${outerColor}, inset 0 0 0 ${
        width + separation
      }px ${innerColor}`;
      break;

    case 'inner':
      boxShadow = `inset 0 0 0 ${separation}px ${innerColor}, 0 0 0 ${width}px ${outerColor}`;
      break;

    default:
      boxShadow = `0 0 0 ${separation}px ${innerColor}, 0 0 0 ${
        width + separation
      }px ${outerColor}`;
      break;
  }

  return {boxShadow} as FocusRingReturn<Width, Separation, InnerColor, OuterColor, Inset>;
}

/**
 * Helper type to handle adding 2 numbers together. Only supports up to 2 + 2 which is all that we
 * see in the wild.
 */
type Add<T extends number, U extends number> = T extends 1
  ? U extends 1
    ? 2
    : U extends 2
    ? 3
    : never
  : T extends 2
  ? U extends 1
    ? 3
    : U extends 2
    ? 4
    : never
  : never;

/**
 * This type is a bit complicated and essentially duplicates the functionality of the runtime. The purpose
 * is to allow static style extraction of the function. For example:
 *
 * ```ts
 * ...focusRing()
 *
 * // type:
 * {
 *   boxShadow: "0 0 0 2px var(--cnvs-base-palette-french-vanilla-100), 0 0 0 4px var(--cnvs-brand-common-focus-outline)";
 * }
 * ```
 */
export type FocusRingReturn<
  Width extends number = 2,
  Separation extends number = 2,
  InnerColor extends string = `var(${typeof base.frenchVanilla100})`,
  OuterColor extends string = `var(${typeof brand.common.focusOutline})`,
  Inset extends 'inner' | 'outer' | 'neither' = 'neither'
> = Inset extends 'inner'
  ? {
      boxShadow: `inset 0 0 0 ${Separation}px ${OuterColor}, inset 0 0 0 ${Add<
        Width,
        Separation
      >}px ${InnerColor}`;
    }
  : Inset extends 'outer'
  ? {
      boxShadow: `inset 0 0 0 ${Separation}px ${OuterColor}, inset 0 0 0 ${Width}px ${InnerColor}`;
    }
  : {
      boxShadow: `0 0 0 ${Separation}px ${InnerColor}, 0 0 0 ${Add<
        Width,
        Separation
      >}px ${OuterColor}`;
    };
