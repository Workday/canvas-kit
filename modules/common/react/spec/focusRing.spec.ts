import {memoizedFocusRing} from '../lib/styles/focusRing';

describe('memoizedFocusRing', () => {
  test('returns same object (reference equality)', () => {
    const args = {
      ringWidth: 3,
      separationWidth: 1,
      animate: true,
      inset: true,
      innerShadowColor: '#333333',
      outerShadowColor: '#666666',
    };
    const received0 = memoizedFocusRing(args);
    const received1 = memoizedFocusRing(args);
    (memoizedFocusRing.cache as Map<
      string,
      [number, number, string, string, boolean, boolean]
    >).clear();
    const received2 = memoizedFocusRing(args);

    expect(received0).toBe(received1);
    expect(received0).toEqual(received2);
    expect(received0).not.toBe(received2);
  });
});
