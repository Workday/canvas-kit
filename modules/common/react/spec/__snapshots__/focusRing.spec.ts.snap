// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`focusRing return correct animation and boxShadow 1`] = `
Object {
  "animation": "animation-10dvi4o 100ms",
  "boxShadow": "0 0 0 1px #333333, 0 0 0 4px #666666",
}
`;

exports[`focusRing returns correct animation and inset box shadow 1`] = `
Object {
  "animation": "animation-2sf17f 100ms",
  "boxShadow": "inset 0 0 0 4px #666666",
}
`;

exports[`focusRing returns default animation and box shadow 1`] = `
Object {
  "animation": "animation-os2b9x 100ms",
  "boxShadow": "0 0 0 0px #ffffff, 0 0 0 2px #0875e1",
}
`;

exports[`focusRing returns no animation 1`] = `
Object {
  "boxShadow": "0 0 0 0px #ffffff, 0 0 0 2px #0875e1",
}
`;
